apps/rush-lib/src/logic/PathTree.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * @public
+ */
+export interface IPathTreeNode<T> {
+  /**
+   * The value that exactly matches the current relative path
+   */
+  value: T | undefined;
+  /**
+   * Child nodes by subfolder
+   */
+  children: Map<string, IPathTreeNode<T>> | undefined;
+}
+
+/**
+ * This class is used to associate POSIX relative paths, such as those returned by `git` commands,
+ * with entities that correspond with ancestor folders, such as Rush Projects
+ */
+export class PathTree<T> {
+  /**
+   * The root node of the tree, corresponding to the path ''
+   */
+  public readonly root: IPathTreeNode<T>;
+
+  /**
+   * Constructs a new `PathTree`
+   *
+   * @param entries - Initial path-value pairs to populate the tree.
+   */
+  public constructor(entries?: Iterable<[string, T]>) {
+    this.root = {
+      value: undefined,
+      children: undefined
+    };
+
+    if (entries) {
+      for (const [path, item] of entries) {
+        this.set(path, item);
+      }
+    }
+  }
+
+  /**
+   * Iterates over the segments of a posix relative path.
+   *
+   * @example
+   * `PathTree.iteratePathSegments('foo/bar/baz')` yields 'foo', 'bar', 'baz'
+   */
+  public static *iteratePathSegments(posixRelativePath: string): Iterable<string> {
```

Comment:
The `yield` pattern has a fair amount of overhead for setting up a generator and managing its state.  This price is worthwhile if we need to:
- Avoid storing the entire sequence in memory; OR
- Avoid calculating items if they are not actually consumed; OR
- Distribute the cost of calculating items across each iteration of a loop

None of those benefits seem to apply here, given that file paths typically have relatively few segments (e.g. <100).  Would it be more efficient to return a simple array?

Reply:
This is, in fact, used to do all three of these things. We only need the first 3 or so elements in the sequence, and we prefer not to allocate unnecessary arrays.
Benchmarking shows that the file path association of git ls-tree to the projects in the test monorepo has:
~54 ms using the generator
~70 ms using `.split('/')`